### PR TITLE
Feat/498 create api to fetch client search paginated screen

### LINKED ISSUE #498: Create API to fetch Client Search paginated screen
Create an endpoint to fetch Forest Client with a pagination option.

Consider these fields on the search - client or agency:
- Name
- Acronym
- Number

Right now it is not possible to fetch data that contains the string the user is trying to filter.

Subtask of #447 
Depends on #525

### PR DESCRIPTION
**Pull request by craigyu**
# Description
Closes #498

Shoutout to Paulo Cruz for pushing the x-total-count feature to prod and helping me solving a super annoying issue with Spring boot, basically that x-total-count response header does not show up in the search by name endpoint and only that endpoint. We didn't find a solution but found a different way of implementation that worked.

From Paulo for future reference(if needed):
> I still don't know why it is behaving like this, but based on some log messages I got, it could be that the resttemplate is dropping the connection before our reactive endpoint had a chance to reply. But I could be completely wrong as the code that worked has no major difference with the one you're using, apart from the fact that is using templating for the URL itself. I'll keep digging later to see if I can come up with a better answer, but so far, it should work with just some small code adjustments.

#### New
- image 
- This endpoint utilizes the fetch by identity logic that is already in place, but it still asks consumers to specify the type of their query. I think this offers more flexibility for future implementation. 
- The tests are mostly there for code coverage. I've done some manual tests with different type of query to see if I get back the correct result, and so far they are correct. You could test with acronym: mof(I also made the search case insensitive for acronym), client_name: ministry of forests, client_number: 00012797 (this one has 66 locations)

#### Changed
- `/api/forest-clients/{clientNumber}/locations` can fetch either 10 (default) or all locations with a client number

#### Removed
- None

### How was this tested?
- [ ] 🧠 Not needed
- [x] 👀 Eyeball
- [x] 🤖 Added tests

###  What gif/image best describes this PR or how it makes you feel?

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Frontend](https://nr-spar-8-frontend.apps.silver.devops.gov.bc.ca/)
[Backend](https://nr-spar-8-backend.apps.silver.devops.gov.bc.ca/actuator/health)
[Oracle-API](https://nr-spar-8-oracle-api.apps.silver.devops.gov.bc.ca/actuator/health)

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-spar/actions/workflows/merge-main.yml)

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Frontend](https://nr-spar-8-frontend.apps.silver.devops.gov.bc.ca/)
[Backend](https://nr-spar-8-backend.apps.silver.devops.gov.bc.ca/actuator/health)
[Oracle-API](https://nr-spar-8-oracle-api.apps.silver.devops.gov.bc.ca/actuator/health)

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-spar/actions/workflows/merge-main.yml)

### CODE REVIEW

**Review by mgaseta — APPROVED** [review]
LGTM!
Nice job!

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java [hunk 1]
```diff
@@ -97,8 +100,14 @@ public List<ForestClientLocationDto> fetchClientLocations(
...
               description = "Number that identifies the client to get the locations.",
               required = true)
-          String clientNumber) {
-    return forestClientService.fetchClientLocations(clientNumber);
+          String clientNumber,
+      @RequestParam(defaultValue = "false")
+          @Parameter(
+              name = "shouldFetchAll",
+              in = ParameterIn.QUERY,
+              description = "If it should fetch all client locations.")
```
Comment (RMCampos) [inline backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java:108]: suggestion: add `required = false` if this is not required.

File: backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java [hunk 2]
```diff
@@ -43,9 +47,11 @@ public Optional<ForestClientDto> fetchClient(String identifier) {
    * @param clientNumber the ForestClient identifier to fetch their locations
    * @return a list of {@link ForestClientLocationDto} containing the client's locations data
    */
-  public List<ForestClientLocationDto> fetchClientLocations(String clientNumber) {
-    SparLog.info("Fetching up to 10 first Forest Clients by its number {}", clientNumber);
-    return forestClientApiProvider.fetchLocationsByClientNumber(clientNumber);
+  public List<ForestClientLocationDto> fetchClientLocations(
+      String clientNumber, Boolean shouldFetchAll) {
+    SparLog.info(
+        "Fetching {} Forest Clients by its number {}", clientNumber, shouldFetchAll ? "all" : "10");
```
Comment (RMCampos) [inline backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java:53]: issue: `clientNumber` should switch place with `shouldFetchAll`. They are in the wrong order.

File: backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java [hunk 3]
```diff
@@ -63,4 +69,103 @@ public ForestClientLocationDto fetchSingleClientLocation(
...
+
+  /**
+   * Search for clients by either the acronym, client number or client name.
+   *
+   * @param type the type of the search query: acronym, client number or client name.
+   * @param query the keyword to search
+   * @return a list of {@link ForestClientSearchDto}
+   */
+  public List<ForestClientSearchDto> searchClients(String type, String query) {
+    if (!type.equals("acronym") && !type.equals("client_number") && !type.equals("client_name")) {
```
Comment (RMCampos) [inline backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java:81]: suggestion: you can have a more readable condition here, something like `if (!List.of("acronym", "client_number", "client_name").contains(type)) { ... }`

File: backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java [hunk 4]
```diff
@@ -63,4 +69,103 @@ public ForestClientLocationDto fetchSingleClientLocation(
...
+
+    clientsFound.forEach(
+        client -> {
+          List<ForestClientLocationDto> fcLocations =
+              forestClientApiProvider.fetchLocationsByClientNumber(client.clientNumber(), true);
+          fcLocations.forEach(
+              fcl -> {
+                mappedList.add(combineClientAndLocationRecord(client, fcl));
+              });
+        });
```
Comment (RMCampos) [inline backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java:109]: praise: nice 👍🏻

File: backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java [hunk 5]
```diff
@@ -63,4 +69,103 @@ public ForestClientLocationDto fetchSingleClientLocation(
...
+
+    List<ForestClientLocationDto> fcLocations =
+        forestClientApiProvider.fetchLocationsByClientNumber(clientNumber, true);
+
+    List<ForestClientSearchDto> mappedList = new ArrayList<>();
+
+    fcLocations.forEach(
+        fcl -> {
+          mappedList.add(combineClientAndLocationRecord(fc, fcl));
+        });
```
Comment (RMCampos) [inline backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java:136]: praise: nice 👍🏻

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java [hunk 6]
```diff
@@ -135,4 +144,29 @@ public ForestClientLocationDto fetchSingleClientLocation(
...
+          @Parameter(
+              name = "type",
+              in = ParameterIn.QUERY,
+              description =
+                  "Specify one of the search types: acronym, client_number or client_name.")
+          String type,
+      @RequestParam
+          @Size(max = 100, message = "Parameter 'query' should be at most 100 characters long.")
+          @Parameter(name = "query", in = ParameterIn.QUERY, description = "The search keyword.")
+          String query) {
```
Comment (RMCampos) [inline backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java:169]: suggestion: same here. Add required true or false.

**Review by RMCampos — CHANGES_REQUESTED** [review]
Great job! 🚀 Only a few suggestions and issues.

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java (hunk 1 above)
Comment (craigyu) [reply backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java:108]: i had that before, then I read that it default to required so I removed it, but yes I agree having it make it more readable.

File: backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java (hunk 2 above)
Comment (craigyu) [reply backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java:53]: oops

Comment (craigyu) [reply backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java]: https://github.com/bcgov/nr-spar/pull/808/commits/4c3c08f2c6313a2de8d97d61e9b8bce0367999d3

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java (hunk 6 above)
Comment (craigyu) [reply backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java:175]: https://github.com/bcgov/nr-spar/pull/808/commits/823d5abb9bc99e9b3dc92f060a2712ddfef7f7a7

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java (hunk 1 above)
Comment (craigyu) [reply backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java]: updated: https://github.com/bcgov/nr-spar/pull/808/commits/823d5abb9bc99e9b3dc92f060a2712ddfef7f7a7

File: backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java (hunk 3 above)
Comment (craigyu) [reply backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java]: good idea: https://github.com/bcgov/nr-spar/pull/808/commits/7f1b58e44721d620439620a5569d85f2aa49eb30

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java (hunk 6 above)
Comment (RMCampos) [reply backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java:175]: Thank you!

**Review by RMCampos — APPROVED** [review]
LGTM 🚀

File: backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java (hunk 1 above)
Comment (RMCampos) [reply backend/src/main/java/ca/bc/gov/backendstartapi/endpoint/ForestClientEndpoint.java]: Thank you!

File: backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java (hunk 2 above)
Comment (RMCampos) [reply backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java]: Thank you!

File: backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java (hunk 3 above)
Comment (RMCampos) [reply backend/src/main/java/ca/bc/gov/backendstartapi/service/ForestClientService.java]: Thank you!

**Pull request merged**